[GTK] Create WebKitWebExtensionMatchPattern

ItsJamie9494 opened the pull request:
#### b2a4be47a0aa5938e31bb00b5c4ebfa5215077a3
<pre>
[GTK] Create WebKitWebExtensionMatchPattern
<a href="https://webkit.org/b/284216">https://webkit.org/b/284216</a>

Reviewed by Michael Catanzaro, Timothy Hatcher, and Adrian Perez de Castro.

This creates a GLib API of WebExtensionMatchPattern, which will then be used
by GLib APIs for WebExtensionContext and WebExtension itself.

* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/glib/WebKitError.cpp:
(webkit_web_extension_error_quark):
* Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.cpp:
(toWebKitWebExtensionError):
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp: Added.
(_WebKitWebExtensionMatchPattern::_WebKitWebExtensionMatchPattern):
(webkit_web_extension_match_pattern_register_custom_url_scheme):
(webkitWebExtensionMatchPatternCreate):
(webkit_web_extension_match_pattern_ref):
(webkit_web_extension_match_pattern_unref):
(webkit_web_extension_match_pattern_new_all_urls):
(webkit_web_extension_match_pattern_new_all_hosts_and_schemes):
(webkit_web_extension_match_pattern_new_with_string):
(webkit_web_extension_match_pattern_new_with_scheme):
(webkit_web_extension_match_pattern_get_string):
(webkit_web_extension_match_pattern_get_scheme):
(webkit_web_extension_match_pattern_get_host):
(webkit_web_extension_match_pattern_get_path):
(webkit_web_extension_match_pattern_get_matches_all_urls):
(webkit_web_extension_match_pattern_get_matches_all_hosts):
(toImpl):
(webkit_web_extension_match_pattern_matches_url):
(webkit_web_extension_match_pattern_matches_pattern):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPatternPrivate.h: Added.
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebExtensionMatchPattern.cpp: Added.
(toPattern):
(testPatternValidity):
(testMatchesPattern):
(testMatchesURL):
(testMatchesAllHosts):
(testMatchesAllURLs):
(testCustomURLScheme):
(beforeAll):
(afterAll):

Canonical link: <a href="https://commits.webkit.org/288424@main">https://commits.webkit.org/288424@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/338da7183589a5a96b15f0f3d912044a64200a58

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82438 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2102 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36530 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87570 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33499 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84543 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2173 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9990 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64254 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/service-worker/redirected-response.https.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22008 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85507 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1548 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75006 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44531 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/81891 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1446 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29192 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32540 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72738 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29827 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88927 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9744 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7006 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72656 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9970 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70820 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71873 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17944 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16012 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15018 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/1120 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9697 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15218 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9571 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13037 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11341 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
```diff
@@ -0,0 +1,353 @@
+ * Returns: (transfer full) (nullable): a newly created #WebKitWebExtensionMatchPattern, or %NULL
+ * if any of the pattern strings are invalid.
+ */
+WebKitWebExtensionMatchPattern* webkit_web_extension_match_pattern_new_with_scheme(const gchar* scheme, const gchar* host, const gchar* path, GError** error)
```
Maybe this instead?
```suggestion
WebKitWebExtensionMatchPattern* webkit_web_extension_match_pattern_new_with_parts(const gchar* scheme, const gchar* host, const gchar* path, GError** error)
```

xeenon commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
```diff
@@ -0,0 +1,353 @@
+ * 
+ * Returns: Whether the pattern matches the specified URL.
+ */
+gboolean webkit_web_extension_match_pattern_matches_url(WebKitWebExtensionMatchPattern* matchPattern, const gchar* url, WebKitWebExtensionMatchPatternOptions  options)
```
```suggestion
gboolean webkit_web_extension_match_pattern_matches_url(WebKitWebExtensionMatchPattern* matchPattern, const gchar* url, WebKitWebExtensionMatchPatternOptions options)
```

xeenon commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
```diff
@@ -0,0 +1,353 @@
+
+    if (options & WEBKIT_WEB_EXTENSION_MATCH_PATTERN_OPTIONS_MATCH_BIDIRECTIONALLY)
+        g_warning("Invalid parameter: WEBKIT_WEB_EXTENSION_MATCH_PATTERN_OPTIONS_MATCH_BIDIRECTIONALLY is not valid when matching a URL");
+
+
```
```suggestion

```

xeenon commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
```diff
@@ -0,0 +1,353 @@
+ * 
+ * Returns: Whether the pattern matches the specified @pattern.
+ */
+gboolean webkit_web_extension_match_pattern_matches_pattern(WebKitWebExtensionMatchPattern* matchPattern, WebKitWebExtensionMatchPattern* pattern, WebKitWebExtensionMatchPatternOptions  options)
```
```suggestion
gboolean webkit_web_extension_match_pattern_matches_pattern(WebKitWebExtensionMatchPattern* matchPattern, WebKitWebExtensionMatchPattern* pattern, WebKitWebExtensionMatchPatternOptions options)
```

xeenon replied:
```suggestion
gboolean webkit_web_extension_match_pattern_matches_pattern(WebKitWebExtensionMatchPattern* matchPattern, WebKitWebExtensionMatchPattern* pattern, WebKitWebExtensionMatchPatternOptions options)
```

xeenon commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
```diff
@@ -0,0 +1,353 @@
+ * This method should be used to register any custom URL schemes used by the app for the extension base URLs,
+ * other than `webkit-extension`, or if extensions should have access to other supported URL schemes when using `<all_urls>`.
+ */
+void webkit_web_extension_match_pattern_register_custom_URL_scheme(const gchar* urlScheme)
```
```suggestion
void webkit_web_extension_match_pattern_register_custom_url_scheme(const gchar* urlScheme)
```

xeenon reviewed (approved):
Looks good. You should get a GTK reviewer for the platform details.

mcatanzaro commented on Source/WebKit/PlatformGTK.cmake:
```diff
@@ -389,6 +390,7 @@ GENERATE_GLIB_API_HEADERS(WebKit WebKitGTK_HEADER_TEMPLATES
     "-DWTF_PLATFORM_GTK=1"
     "-DWTF_PLATFORM_WPE=0"
     "-DUSE_GTK4=$<BOOL:${USE_GTK4}>"
+    "-DENABLE_WK_WEB_EXTENSIONS=$<BOOL:${ENABLE_WK_WEB_EXTENSIONS}>"
```
Nooo don't do this! Having conditionally-compiled public APIs will be confusing and not fun. Let's keep things fun. :)

What we should do instead is always build all of the APIs, and leave them as stubs in case `ENABLE_WK_WEB_EXTENSIONS` is not defined. Generally you would return a default value (e.g. 0 or empty string) from functions that return values, and otherwise do nothing. You can use `ENABLE_WK_WEB_EXTENSIONS` in the source code, but not in the public headers. Since you won't use it in the headers, you don't need to pass it to unifdef here.

(We should also consider -- separately -- whether `ENABLE_WK_WEB_EXTENSIONS` is really needed at all. Unless it's necessary to guard use of some external dependency on Apple platforms, then it can probably be removed in favor of using a runtime feature flag.)

xeenon replied:
We need the define for Apple platforms.

ItsJamie9494 replied:
If the define is needed, then I can add stub functions. I don't love having a useless API in builds, especially if it's not immediately clear that it's useless, but if it's done elsewhere that's fine.

mcatanzaro replied:
It's not useless, it's _conditionally_ useless. :) Conditional existence is a much bigger pain (are the APIs missing from the docs? does your application fail to build depending on WebKitGTK build flags?). Yes it's done elsewhere, although generally for compatibility purposes; we probably won't want to actually ship a stable release of WebKitGTK with stub APIs by default, so if we're not ready to enable WebExtensions soon then we might want to revert this work on our next stable branch.

mcatanzaro replied:
You could also just enable the setting (`WEBKIT_OPTION_DEFAULT_PORT_VALUE` in OptionsGTK.cmake and possibly OptionsWPE.cmake) if ready to do so.

ItsJamie9494 replied:
I'd rather not enable the setting just yet, since there's still a bit more work to make it useful. I do hope to ship this soon, or at least ship some of the APIs, so I'm hoping we don't have to revert this for a stable branch.

ItsJamie9494 replied:
it would be nice if I could remmeber why I made a change.

these lines in PlatformGTK and PlatformWPE weren't for enabling the API, it was for properly creating the error enumeration types (or not, if the API was disabled). The points about conditional API creation are valid, but without these lines, the error enumerations have to always exist as glib-mkenums doesn't recognise that they're conditionally enabled.

mcatanzaro replied:
You'll need to make the error enumerations always exist. That seems harmless enough.

aperezdc replied:
I'm also not a fan of having the symbols (and header) enabled or not depending on a build time setting. We should decide if we want the feature, and act accordingly. We are landing this on `main`, which is where in-development features happen, so IMO it's fine to have it unconditionally enabled.

Later on, when doing the next stable branch, we can decide to remove the public API (also unconditionally) _only in the branch_ if the feature is not considered ready for public consumption.

ItsJamie9494 replied:
I had fixed this on a local branch, where the enumerations this line enabled are just always enabled. Apparently I forgot to push it though.

This means we're not enabling things with build-time settings in cmake (although the WK_WEB_EXTENSIONS feature still exists and needs to keep existing for Apple platforms. Right now it's not enabled on Linux, but that's just a few lines in wtf/PlatformGtk i believe)

mcatanzaro reviewed (changes requested):
Since this adds new public API, we actually need two GTK/WPE reviewers to approve this one. Either Adrian or Carlos Garcia would be a good bet for a second reviewer.

I just took a quick look for now and have one big request. Will look closer once you've responded to that.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
```diff
@@ -27,22 +27,28 @@
+#define WEBKIT_SNAPSHOT_ERROR                    webkit_snapshot_error_quark ()
+#define WEBKIT_USER_CONTENT_FILTER_ERROR         webkit_user_content_filter_error_quark ()
+
+#define WEBKIT_WEB_EXTENSION_MATCH_PATTERN_ERROR webkit_web_extension_error_quark ()
```
Nit: these were alphabetized!

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitPrivate.cpp:
```diff
@@ -150,6 +154,23 @@ unsigned toWebKitError(unsigned webCoreError)
+        return WEBKIT_WEB_EXTENSION_MATCH_PATTERN_ERROR_INVALID_HOST;
+    case static_cast<int>(WebKit::WebExtensionMatchPattern::Error::InvalidPath):
+        return WEBKIT_WEB_EXTENSION_MATCH_PATTERN_ERROR_INVALID_PATH;
+    case static_cast<int>(WebKit::WebExtensionMatchPattern::Error::Unknown):
```
Nite: might be nicer to cast apiError to `WebKit::WebExtensionMatchPattern::Error` once instead of casting to int four times.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
```diff
@@ -0,0 +1,441 @@
+/**
+ * WebKitWebExtensionMatchPattern: (ref-func webkit_web_extension_match_pattern_ref) (unref-func webkit_web_extension_match_pattern_unref)
+ * 
+ * Represents a way to specify a group of URLs for use in Web Extensions.
```
Should this be WebExtensions?

```suggestion
 * Represents a way to specify a group of URLs for use in WebExtensions.
```

Or possibly "web extensions"? Probably not "Web Extensions."

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
```diff
@@ -0,0 +1,441 @@
+ */
+struct _WebKitWebExtensionMatchPattern {
+#if ENABLE(WK_WEB_EXTENSIONS)
+    _WebKitWebExtensionMatchPattern(RefPtr<WebExtensionMatchPattern>&& matchPattern)
```
```suggestion
    explicit _WebKitWebExtensionMatchPattern(RefPtr<WebExtensionMatchPattern>&& matchPattern)
```

ItsJamie9494 replied:
interestingly, this causes a style checker failure about _WebExtensionMatchPattern being incorrectly named. It's done in other places in WebKit/the gtk api, so I'm not sure why it's denying it here.

mcatanzaro replied:
Weird. You can add another style checker exception, or ignore it.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
```diff
@@ -0,0 +1,441 @@
+ * consist of three parts: scheme, host, and path.
+ * 
+ * Generally, match patterns are returned from a #WebKitWebExtension.
+ */
```
```suggestion
 * Generally, match patterns are returned from a #WebKitWebExtension.
 *
 * Since: 2.48
 */
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
```diff
@@ -0,0 +1,441 @@
+ * 
+ * This method should be used to register any custom URL schemes used by the app for the extension base URLs,
+ * other than `webkit-extension`, or if extensions should have access to other supported URL schemes when using `<all_urls>`.
+ */
```
```suggestion
 * other than `webkit-extension`, or if extensions should have access to other supported URL schemes when using `<all_urls>`.
 *
 * Since: 2.48
 */
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
```diff
@@ -0,0 +1,441 @@
+    WebKitWebExtensionMatchPattern* pattern = static_cast<WebKitWebExtensionMatchPattern*>(fastMalloc(sizeof(WebKitWebExtensionMatchPattern)));
+    new (pattern) WebKitWebExtensionMatchPattern(static_pointer_cast<WebExtensionMatchPattern>(apiMatchPattern));
+    return pattern;
+}
```
Is it really necessary to have separate constructors for RefPtr and Ref?

If you do need to keep it, then at least implement the Ref constructor using the RefPtr constructor instead of copying all of this twice, to avoid inconsistencies. E.g. you have an assert in the RefPtr version that is missing here.

ItsJamie9494 replied:
Yes, WebExtensionMatchPattern itself returns RefPtr's, and WebExtension has some functions that return Ref, so when we expose a Gtk API for WebExtension, Ref will be needed. For the Ref implementation, I can certainly just make it call the RefPtr constructor, that's fine

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
```diff
@@ -0,0 +1,441 @@
+ * This function is MT-safe and may be called from any thread.
+ *
+ * Returns: The same @matchPattern with an additional reference.
+ */
```
```suggestion
 * Returns: The same @matchPattern with an additional reference.
 *
 * Since: 2.48
 */
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
```diff
@@ -0,0 +1,441 @@
+ * If the reference was the last, the resources associated to the
+ * @matchPattern are freed. This function is MT-safe and may be called from
+ * any thread.
+ */
```
```suggestion
 * any thread.
 *
 * Since: 2.48
 */
```

I will stop adding suggestions for the Since tags. Just add them everywhere please!

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp:
```diff
@@ -0,0 +1,441 @@
+    }
+
+    RefPtr<WebExtensionMatchPattern> matchPattern;
+    CString string { matchPattern->string().utf8() };
+    CString scheme { matchPattern->scheme().utf8() };
+    CString host { matchPattern->host().utf8() };
+    CString path { matchPattern->path().utf8() };
+    bool matchesAllURLs { matchPattern->matchesAllURLs() };
+    bool matchesAllHosts { matchPattern->matchesAllHosts() };
```
Our API structs normally initialize data like this lazily only when needed, but I think your way is fine too. Certainly this is simpler, and I do like simple.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.h.in:
```diff
@@ -0,0 +1,111 @@
+ * @WEBKIT_WEB_EXTENSION_MATCH_PATTERN_OPTIONS_MATCH_BIDIRECTIONALLY: Two patterns should be checked in either direction while matching (A matches B, or B matches A). Invalid for matching URLs.
+ *
+ * Enum values representing matching options.
+ */
```
```suggestion
 * Enum values representing matching options.
 *
 * Since: 2.48
 */
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.h.in:
```diff
@@ -0,0 +1,111 @@
+    WEBKIT_WEB_EXTENSION_MATCH_PATTERN_OPTIONS_MATCH_BIDIRECTIONALLY = 1 << 3,
+} WebKitWebExtensionMatchPatternOptions;
+
+#define WEBKIT_TYPE_WEB_EXTENSION_MATCH_PATTERN            (webkit_web_extension_match_pattern_get_type())
```
```suggestion
#define WEBKIT_TYPE_WEB_EXTENSION_MATCH_PATTERN            (webkit_web_extension_match_pattern_get_type ())
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.h.in:
```diff
@@ -0,0 +1,111 @@
+typedef struct _WebKitWebExtensionMatchPattern WebKitWebExtensionMatchPattern;
+
+WEBKIT_API GType
+webkit_web_extension_match_pattern_get_type (void) G_GNUC_CONST;
```
We have a debate in the comments of https://blogs.gnome.org/mcatanzaro/2015/09/14/your-_get_type-function-is-not-g_gnuc_const/ on whether to use `G_GNUC_CONST` here or not. I don't like it and currently WebKit only does so in WebKitFeature.h.in, so probably shouldn't.

mcatanzaro commented on Tools/Scripts/webkitpy/style/checker.py:
```diff
@@ -199,6 +199,11 @@
+    ([
+        # The MatchPattern API uses a lowercase "url" to match GObject standard naming schemes
+        os.path.join('Source', 'WebKit', 'UIProcess', 'API', 'glib', 'WebKitWebExtensionMatchPattern.cpp')],
+     ["-readability/naming/acronym"]),
```
Nice! Many developers just ignore style errors instead of updating the style checker like you've done here.

mcatanzaro reviewed (changes requested):
Well, this looks pretty much good to me.

I also like the large number of new tests, except I think they're not actually built? You need to add it using ADD_WK2_TEST() in Tools/TestWebKitAPI/glib/CMakeLists.txt.

mcatanzaro commented on Tools/TestWebKitAPI/glib/CMakeLists.txt:
```diff
@@ -180,6 +180,7 @@ ADD_WK2_TEST(TestWebKitSecurityOrigin ${TOOLS_DIR}/TestWebKitAPI/Tests/WebKitGLi
 ADD_WK2_TEST(TestWebKitSettings ${TOOLS_DIR}/TestWebKitAPI/Tests/WebKitGLib/TestWebKitSettings.cpp)
 ADD_WK2_TEST(TestWebKitURIUtilities ${TOOLS_DIR}/TestWebKitAPI/Tests/WebKitGLib/TestWebKitURIUtilities.cpp)
 ADD_WK2_TEST(TestWebKitWebContext ${TOOLS_DIR}/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebContext.cpp)
+ADD_WK2_TEST(TestWebKitWebExtensionMatchPattern ${TOOLS_DIR}/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebExtensionMatchPattern.cpp)
```
This broke WPE:

```
[2068/2251] Linking CXX executable bin/TestWebKitAPI/WPE/TestWebKitWebExtensionMatchPattern
FAILED: bin/TestWebKitAPI/WPE/TestWebKitWebExtensionMatchPattern 
: && /usr/bin/c++ -DASSERT_ENABLED=1 -fdiagnostics-color=always -Wextra -Wall -pipe -fmax-errors=20 -Wno-odr -Wno-stringop-overread -Wno-stringop-overflow -Wno-nonnull -Wno-array-bounds -Wno-expansion-to-defined -Wno-noexcept-type -Wno-psabi -Wno-misleading-indentation -Wno-maybe-uninitialized -Wundef -Wpointer-arith -Wmissing-format-attribute -Wformat-security -Wcast-align -Wno-tautological-compare -fasynchronous-unwind-tables -fno-strict-aliasing -fno-exceptions -fno-rtti -fcoroutines -ffunction-sections -fdata-sections -O3 -DNDEBUG -fuse-ld=lld -Wl,--gc-sections -Wl,--disable-new-dtags Tools/TestWebKitAPI/glib/CMakeFiles/TestWebKitWebExtensionMatchPattern.dir/__/Tests/WebKitGLib/TestWebKitWebExtensionMatchPattern.cpp.o -o bin/TestWebKitAPI/WPE/TestWebKitWebExtensionMatchPattern  -Wl,-rpath,/app/webkit/WebKitBuild/WPE/Release/lib  /usr/lib/x86_64-linux-gnu/libglib-2.0.so  /usr/lib/x86_64-linux-gnu/libsoup-3.0.so  lib/libWebKitGLibAPITestsCore.a  lib/libWPEWebKit-2.0.so.2.4.1  /usr/lib/x86_64-linux-gnu/libseccomp.so  /usr/lib/x86_64-linux-gnu/libicudata.so  /usr/lib/x86_64-linux-gnu/libicui18n.so  /usr/lib/x86_64-linux-gnu/libicuuc.so  /usr/lib/x86_64-linux-gnu/libsystemd.so  /usr/lib/x86_64-linux-gnu/libbacktrace.a  /usr/lib/x86_64-linux-gnu/libsysprof-capture-4.a  /usr/lib/x86_64-linux-gnu/libxml2.so  /usr/lib/x86_64-linux-gnu/libsqlite3.so  /usr/lib/x86_64-linux-gnu/libxslt.so  lib/libGLESv2.a  lib/libANGLE.a  /usr/lib/x86_64-linux-gnu/liblcms2.so  /usr/lib/x86_64-linux-gnu/libwoff2dec.so  lib/libxdgmime.a  /usr/lib/x86_64-linux-gnu/libharfbuzz-icu.so  /usr/lib/x86_64-linux-gnu/libgcrypt.so  /usr/lib/x86_64-linux-gnu/libgpg-error.so  /usr/lib/x86_64-linux-gnu/libcrypto.so  /usr/lib/x86_64-linux-gnu/libwebpdemux.so  /usr/lib/x86_64-linux-gnu/libjxl.so  /usr/lib/x86_64-linux-gnu/libavif.so  -latomic  /usr/lib/x86_64-linux-gnu/libgio-2.0.so  /usr/lib/x86_64-linux-gnu/libgstallocators-1.0.so  /usr/lib/x86_64-linux-gnu/libgstapp-1.0.so  /usr/lib/x86_64-linux-gnu/libgstbase-1.0.so  /usr/lib/x86_64-linux-gnu/libgstpbutils-1.0.so  /usr/lib/x86_64-linux-gnu/libgstaudio-1.0.so  /usr/lib/x86_64-linux-gnu/libgsttranscoder-1.0.so  /usr/lib/x86_64-linux-gnu/libgstgl-1.0.so  /usr/lib/x86_64-linux-gnu/libgsttag-1.0.so  /usr/lib/x86_64-linux-gnu/libgstvideo-1.0.so  /usr/lib/x86_64-linux-gnu/libgstrtp-1.0.so  /usr/lib/x86_64-linux-gnu/libgstsdp-1.0.so  /usr/lib/x86_64-linux-gnu/libgstwebrtc-1.0.so  /usr/lib/x86_64-linux-gnu/libgstfft-1.0.so  /usr/lib/x86_64-linux-gnu/libocdm.so  /usr/lib/x86_64-linux-gnu/libflite.so  /usr/lib/x86_64-linux-gnu/libflite_usenglish.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_grapheme_lang.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_grapheme_lex.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_indic_lang.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_indic_lex.so  /usr/lib/x86_64-linux-gnu/libflite_cmulex.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_time_awb.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_us_awb.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_us_kal16.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_us_kal.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_us_rms.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_us_slt.so  /usr/lib/x86_64-linux-gnu/libtasn1.so  /usr/lib/x86_64-linux-gnu/libopenxr_loader.so  -latomic  /usr/lib/x86_64-linux-gnu/libgio-2.0.so  /usr/lib/x86_64-linux-gnu/libgstallocators-1.0.so  /usr/lib/x86_64-linux-gnu/libgstapp-1.0.so  /usr/lib/x86_64-linux-gnu/libgstbase-1.0.so  /usr/lib/x86_64-linux-gnu/libgstpbutils-1.0.so  /usr/lib/x86_64-linux-gnu/libgstaudio-1.0.so  /usr/lib/x86_64-linux-gnu/libgsttranscoder-1.0.so  /usr/lib/x86_64-linux-gnu/libgstgl-1.0.so  /usr/lib/x86_64-linux-gnu/libgsttag-1.0.so  /usr/lib/x86_64-linux-gnu/libgstvideo-1.0.so  /usr/lib/x86_64-linux-gnu/libgstrtp-1.0.so  /usr/lib/x86_64-linux-gnu/libgstsdp-1.0.so  /usr/lib/x86_64-linux-gnu/libgstwebrtc-1.0.so  /usr/lib/x86_64-linux-gnu/libgstfft-1.0.so  /usr/lib/x86_64-linux-gnu/libocdm.so  /usr/lib/x86_64-linux-gnu/libflite.so  /usr/lib/x86_64-linux-gnu/libflite_usenglish.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_grapheme_lang.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_grapheme_lex.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_indic_lang.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_indic_lex.so  /usr/lib/x86_64-linux-gnu/libflite_cmulex.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_time_awb.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_us_awb.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_us_kal16.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_us_kal.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_us_rms.so  /usr/lib/x86_64-linux-gnu/libflite_cmu_us_slt.so  /usr/lib/x86_64-linux-gnu/libtasn1.so  /usr/lib/x86_64-linux-gnu/libopenxr_loader.so  /usr/lib/x86_64-linux-gnu/libgmodule-2.0.so  /usr/lib/x86_64-linux-gnu/libsoup-3.0.so  /usr/lib/x86_64-linux-gnu/libgstreamer-1.0.so  -ldl  /usr/lib/x86_64-linux-gnu/libgbm.so  /usr/lib/x86_64-linux-gnu/libdrm.so  lib/libWPEToolingBackends.a  /usr/lib/x86_64-linux-gnu/libgobject-2.0.so  /usr/lib/x86_64-linux-gnu/libglib-2.0.so  lib/libSkia.a  /usr/lib/x86_64-linux-gnu/libepoxy.so  /usr/lib/x86_64-linux-gnu/libharfbuzz.so  /usr/lib/x86_64-linux-gnu/libjpeg.so  /usr/lib/x86_64-linux-gnu/libpng.so  /usr/lib/x86_64-linux-gnu/libz.so  /usr/lib/x86_64-linux-gnu/libfontconfig.so  /usr/lib/x86_64-linux-gnu/libfreetype.so  /usr/lib/x86_64-linux-gnu/libexpat.so  /usr/lib/x86_64-linux-gnu/libwebpmux.so  /usr/lib/x86_64-linux-gnu/libwebp.so  /usr/lib/x86_64-linux-gnu/libwpe-1.0.so  /usr/lib/x86_64-linux-gnu/libatk-bridge-2.0.so  /usr/lib/x86_64-linux-gnu/libWPEBackend-fdo-1.0.so  -lxkbcommon  -lwayland-server  -lwayland-egl  -lwayland-client  /usr/lib/x86_64-linux-gnu/libatk-1.0.so  -Wl,-rpath-link,/app/webkit/WebKitBuild/WPE/Release/lib && :
ld.lld: error: undefined symbol: beforeAll()
>>> referenced by TestMain.cpp
>>>               ../Tools/TestWebKitAPI/glib/CMakeFiles/WebKitGLibAPITestsCore.dir/WebKitGLib/TestMain.cpp.o:(main) in archive lib/libWebKitGLibAPITestsCore.a
ld.lld: error: undefined symbol: afterAll()
>>> referenced by TestMain.cpp
>>>               ../Tools/TestWebKitAPI/glib/CMakeFiles/WebKitGLibAPITestsCore.dir/WebKitGLib/TestMain.cpp.o:(main) in archive lib/libWebKitGLibAPITestsCore.a
collect2: error: ld returned 1 exit status
```

Problem is `ENABLE(WK_WEB_EXTENSIONS)` is not defined. Two options to handle it: (a) always define `beforeAll()` and `afterAll()`, (b) check `ENABLE_WK_WEB_EXTENSIONS` here in the CMake rules to build the entire test conditionally, as is done for TestWebKitNetworkSession just below.

mcatanzaro reviewed (approved):
This gets my approval. Still need a second GTK/WPE reviewer to approve before landing the new API, though.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
```diff
@@ -157,6 +158,22 @@ typedef enum {
+ * @WEBKIT_WEB_EXTENSION_MATCH_PATTERN_ERROR_INVALID_HOST: The host component was invalid.
+ * @WEBKIT_WEB_EXTENSION_MATCH_PATTERN_ERROR_INVALID_PATH: The path component was invalid.
+ *
+ * Enum values used to denote errors happening when creating a #WebKitWebExtensionMatchPattern
+ */
```
```suggestion
 * Enum values used to denote errors happening when creating a #WebKitWebExtensionMatchPattern
 *
 * Since: 2.48
 */
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitError.cpp:
```diff
@@ -112,6 +112,18 @@ GQuark webkit_snapshot_error_quark()
+ *
+ * Gets the quark for the domain of Web Extension Match Pattern errors.
+ *
+ * Returns: web extension match pattern error domain.
+ */
```
```suggestion
 * Returns: web extension match pattern error domain.
 *
 * Since: 2.48
 */
```

Please double check the rest of this pull request for other missing Since tags.

aperezdc reviewed (approved):
This looks good overall to me as well.

If something, my only “complaint” would be that reading things like “match pattern matches” is a bit cacophonic and results in quite long function names. I wonder if we could simplify the name, or have a better name. Maybe just `WebKitWebExtensionPattern` is fine enough? It wouldn't be the first thing we rename in our public API vs. the internal WebKit name trying to make our API nicer to use or somewhat more consistent with naming using in other GLib/GNOME libraries.

ItsJamie9494 committed:
[GTK] Create WebKitWebExtensionMatchPattern
https://webkit.org/b/284216

Reviewed by Michael Catanzaro, Timothy Hatcher, and Adrian Perez de Castro.

This creates a GLib API of WebExtensionMatchPattern, which will then be used
by GLib APIs for WebExtensionContext and WebExtension itself.

* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/glib/WebKitError.cpp:
(webkit_web_extension_error_quark):
* Source/WebKit/UIProcess/API/glib/WebKitError.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.cpp:
(toWebKitWebExtensionError):
* Source/WebKit/UIProcess/API/glib/WebKitPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.cpp: Added.
(_WebKitWebExtensionMatchPattern::_WebKitWebExtensionMatchPattern):
(webkit_web_extension_match_pattern_register_custom_url_scheme):
(webkitWebExtensionMatchPatternCreate):
(webkit_web_extension_match_pattern_ref):
(webkit_web_extension_match_pattern_unref):
(webkit_web_extension_match_pattern_new_all_urls):
(webkit_web_extension_match_pattern_new_all_hosts_and_schemes):
(webkit_web_extension_match_pattern_new_with_string):
(webkit_web_extension_match_pattern_new_with_scheme):
(webkit_web_extension_match_pattern_get_string):
(webkit_web_extension_match_pattern_get_scheme):
(webkit_web_extension_match_pattern_get_host):
(webkit_web_extension_match_pattern_get_path):
(webkit_web_extension_match_pattern_get_matches_all_urls):
(webkit_web_extension_match_pattern_get_matches_all_hosts):
(toImpl):
(webkit_web_extension_match_pattern_matches_url):
(webkit_web_extension_match_pattern_matches_pattern):
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPattern.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitWebExtensionMatchPatternPrivate.h: Added.
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebExtensionMatchPattern.cpp: Added.
(toPattern):
(testPatternValidity):
(testMatchesPattern):
(testMatchesURL):
(testMatchesAllHosts):
(testMatchesAllURLs):
(testCustomURLScheme):
(beforeAll):
(afterAll):

Canonical link: https://commits.webkit.org/288424@main
